Fixed #27434: Removed an edge case with model validation

matthiask opened the pull request:
When raising a dictionary ValidationError in Model.clean() for fields which are not part of a particular modelform, Django would raise a ValueError because it couldn't assign those errors to a form field.

This change moves errors from non-existing fields to NON_FIELD_ERRORS.

A better idea might be to check whether the field actually exists on the model and prepend the field's `verbose_name` to all error message, but I wanted to solicit some feedback first before moving forward with this patch, especially from @loic :-)

Link to related ticket for easier browsing: https://code.djangoproject.com/ticket/27434

matthiask commented:
Thanks for the comments and the review. I adapted the testcase as suggested.

Regarding dropping errors vs. moving them to `NON_FIELD_ERRORS`: The problem with dropping messages is that this might lead to full_clean raising no error despite the failing model validation in Model.clean. So simply dropping them without adding an alternative error somewhere else is probably out of the question.

charettes commented:
> Regarding dropping errors vs. moving them to NON_FIELD_ERRORS: The problem with dropping messages is that this might lead to full_clean raising no error despite the failing model validation in Model.clean. So simply dropping them without adding an alternative error somewhere else is probably out of the question.

There might be cases where you wouldn't want to raise an exception for a specific field if you knew in `clean()` that the field was excluded. Since `clean()` doesn't offer such an API and that raising an exception returns immediately from `clean()`, possibly preventing other exceptions from being raised later in the body of `clean()` I think it makes more sense to move the error the `NON_FIELD_ERRORS` as well.

charettes commented:
Buildbot, add to whitelist.

loic commented:
I'm on the fence on that one, on one hand the current situation is far from ideal, on the other hand I don't like the proposed loss of metadata: the field name is no longer available and the model's error message may not be intelligible without the context of the field. For example "this field is required" doesn't make sense as a non-field error. And you can't even fix that poor error message at the form layer, since you can't target that error, the field name being gone, and the error code may not be specific enough (e.g. 'invalid' vs 'title_invalid').

Maybe a better approach is to pass `exclude` as a parameter to Model.clean() - we should be able to do that in a somewhat backwards compatible way with inspect.formatargspec() - then `CustomModel.clean()` can decide to not raise this field-specific error, or to raise it differently with a message and code that's applicable for NON_FIELD_ERRORS.

charettes commented:
> Maybe a better approach is to pass exclude as a parameter to Model.clean() - we should be able to do that in a somewhat backwards compatible way with inspect.formatargspec() - then CustomModel.clean(exlude) can decide to not raise this field-specific error, or to raise it differently with a message and code that's applicable for NON_FIELD_ERRORS.

It's too bad it can be done without breaking backward compatibility :)

matthiask commented:
@loic I mostly agree with all points you've raised, but am not sure what I should do now to move this foward. I also like breaking stuff ;-) , but a change with this effect surely has a higher barrier, and I don't want to put in too much needless work if there's an easier way.

(Btw, I hope everything went well on Sunday!)

loic commented:
Hey @matthiask, I suggested using the `inspect` module to check it `self.clean()` accepts an `exclude` kwarg. If it does, we provide it, otherwise we don't.

timgraham commented:
Will you continue this @matthiask?

matthiask commented:
Hi @timgraham, I still intend do finish this, yes.

matthiask commented:
I rebased the branch on top of master and implemented the suggestion. The patch is much smaller in terms of code now.

Do you think it makes sense to deprecate `Model.clean()` without an `exclude` parameter? I'm not sure it's necessary to force everyone to change their `clean()` method signature, but I'll happily leave the decision to you.

timgraham commented:
I don't think it would be wise to continue without a deprecation. If we have to add more parameters some time, the possible signatures would grow more complicated over time.

matthiask commented:
Thanks for the feedback! 

I've updated the PR.

auvipy commented:
@matthiask the alpha is schedule for today or may be tomorrow. It would be great if you could complete it before that and rebased the pr with logical commits

matthiask commented:
I've already added the missing test referenced above. @timgraham do you have any additional feedback?

matthiask commented:
I've rebased the patch on top of current master.

I'm not sure how to proceed here. I'm now thinking about simply using and recommending `clean_fields(self, exclude)` over `clean(self)` for anyone wanting to customize error messages for different forms, which means a simple documentation patch might be enough for this.

OTOH it's still possible to crash Django with code from the documentation, so that's not very nice (tm), but it's an edge case...

Comments / opinions?

auvipy commented on docs/ref/models/instances.txt:
```diff
@@ -251,7 +251,11 @@ validation.
-.. method:: Model.clean()
+.. method:: Model.clean(exclude=exclude)
+
+.. versionchanged:: 1.11
```
2.0

auvipy commented on docs/releases/1.11.txt:
```diff
@@ -353,6 +353,10 @@ Models
 * Added :meth:`.QuerySet.union`, :meth:`~.QuerySet.intersection`, and
   :meth:`~.QuerySet.difference`.
 
+* Added optional support for passing the list of excluded fields into model's
```
move this portion to 2.0

matthiask commented:
Thanks @auvipy for the review!

I'm not interested in pursuing this further. The changes are too many, and my personal itch has been thoroughly scratched by [moving all validation logic to `clean_fields()`](https://github.com/matthiask/feincms3/commit/81e9ffde8fa50b8d8a83053340bda5a47f9ab42c) for now.

I've followed my suggestion and [added a note to the documentation](https://github.com/django/django/pull/8394) documenting this problem for posterity.
